protocol/client.go

```diff
@@ -0,0 +1,203 @@
+package protocol
+
+import (
+	"bytes"
+
+	"github.com/coniks-sys/coniks-go/crypto/sign"
+	"github.com/coniks-sys/coniks-go/crypto/vrf"
+	m "github.com/coniks-sys/coniks-go/merkletree"
+)
+
+type ProofType int
+
+const (
+	InvalidProof ProofType = iota
+	ProofOfAbsence
+	ProofOfInclusion
+)
+
+// ConiksClient stores the current state of a CONIKS client.
+// This includes the current epoch and SignedTreeRoot value,
+// as well as directory's configurations
+// (e.g., wherether the TemporaryBinding extension is being used or not).
+// This struct also stores the verification results including
+// the type of returned proof.
+// The client should create a new ConiksClient instance only once,
+// when she registers her binding with a ConiksDirectory.
+// This ConiksClient instance then will be used to verify
+// the returned response from a ConiksDirectory.
+type ConiksClient struct {
+	CurrentEpoch uint64
+	SavedSTR     []byte
+
+	// verification results
+	ProofType          ProofType
+	VerificationResult ErrorCode
+
+	// extensions settings
+	isUseTBs bool
+	TBs      []*m.TemporaryBinding
+
+	// signing key
+	// TODO: should we pin the signing key and vrf key in the client? see #47
+	signKey sign.PublicKey
+}
+
+func NewConiksClient(savedSTR []byte, isUseTBs bool, signKey sign.PublicKey) *ConiksClient {
+	return &ConiksClient{
+		CurrentEpoch: 0,
+		SavedSTR:     savedSTR,
+		isUseTBs:     isUseTBs,
+		signKey:      signKey,
+	}
+}
+
+// Verify verifies the directory's response of the requested username.
+// The verification results are stored in ProofType and VerificationResult
+// field of ConiksClient struct. The client first checks the VerificationResult
+// to see if the response message is passed or not. If it gets a valid response,
+// (i.e., VerificationResult is Passed), it then needs to check the returned proof
+// is matched with its expectation or not. If the client does not check the type
+// of the returned proof, it means the client could be getting a valid proof of absence,
+// when it expects a proof of inclusion, and as long as it's valid, it's none the wiser.
+// The client can ignore the proof type check, and notify the user the verification
+// result instead if the verification check is not passed.
```

Comment:
> the proof type for each response is really part of the protocol

I think there are some cases where the developer/client needs to know beforehand whether it's expecting a proof of inclusion or not, and in such cases, I agree that the proof type check should be included in the pkg. However, there may have  other cases where the client doesn't know which type of proof it should expect (Key lookup case), then how do we handle these cases?